src/main/java/ch/njol/skript/sections/SecWhile.java

```diff
@@ -18,85 +18,190 @@
 @Examples({
 	"while size of all players < 5:",
-	"\tsend \"More players are needed to begin the adventure\" to all players",
-	"\twait 5 seconds",
+		"\tsend \"More players are needed to begin the adventure\" to all players",
+		"\twait 5 seconds",
 	"",
 	"set {_counter} to 1",
 	"do while {_counter} > 1: # false but will increase {_counter} by 1 then get out",
-	"\tadd 1 to {_counter}",
+		"\tadd 1 to {_counter}",
 	"",
 	"# Be careful when using while loops with conditions that are almost ",
 	"# always true for a long time without using 'wait %timespan%' inside it, ",
 	"# otherwise it will probably hang and crash your server.",
 	"while player is online:",
-	"\tgive player 1 dirt",
-	"\twait 1 second # without using a delay effect the server will crash",
+		"\tgive player 1 dirt",
+		"\twait 1 second # without using a delay effect the server will crash",
+	"",
+	"while all:",
+		"\tplayer is online",
+		"\tplayer's experience < 200",
+	"do:",
+		"\twait 1 second",
+		"\tadd 2 to player's experience",
+	"",
+	"while any:",
+		"\tplayer's tool is {_item}",
+		"\tplayer's offhand tool is {_item}",
+	"do:",
+		"\theal player by 1"
 })
-@Since("2.0, 2.6 (do while)")
+@Since("2.0, 2.6 (do while), INSERT VERSION (all, any)")
 public class SecWhile extends LoopSection {
 
-	static {
-		Skript.registerSection(SecWhile.class, "[:do] while <.+>");
-	}
+	private static final SkriptPattern DO_PATTERN = PatternCompiler.compile("do");
 
-	@SuppressWarnings("NotNullFieldNotInitialized")
-	private Condition condition;
+	private enum WhileState {NORMAL, ANY, DO}
 
-	@Nullable
-	private TriggerItem actualNext;
+	private static final Patterns<WhileState> WHILE_STATE_PATTERNS = new Patterns<>(new Object[][] {
+		{"[:do] while <.+>", WhileState.NORMAL},
+		{"[:do] while [all]", WhileState.NORMAL},
+		{"[:do] while (any|at least one of)", WhileState.ANY},
+		{DO_PATTERN.toString(), WhileState.DO}
+	});
+
+	static {
+		Skript.registerSection(SecWhile.class, WHILE_STATE_PATTERNS.getPatterns());
+	}
 
+	private @Nullable TriggerItem actualNext;
 	private boolean doWhile;
 	private boolean ranDoWhile = false;
+	private WhileState state;
+	private boolean multiline;
+	private @UnknownNullability Conditional<Event> conditional;
 
 	@Override
-	public boolean init(Expression<?>[] exprs,
-						int matchedPattern,
-						Kleenean isDelayed,
-						ParseResult parseResult,
-						SectionNode sectionNode,
-						List<TriggerItem> triggerItems) {
-		String expr = parseResult.regexes.get(0).group();
-
-		condition = Condition.parse(expr, "Can't understand this condition: " + expr);
-		if (condition == null)
-			return false;
-
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult, SectionNode sectionNode, List<TriggerItem> triggerItems) {
+		state = WHILE_STATE_PATTERNS.getInfo(matchedPattern);
 		doWhile = parseResult.hasTag("do");
+		multiline = parseResult.regexes.isEmpty() && state != WhileState.DO;
```

Comment:
```suggestion
		multiline = matchedPattern == 1 || matchedPattern == 2;
```
I feel like this is a bit easier to understand